src/vs/base/common/lifecycle.ts

```diff
@@ -108,22 +151,39 @@ export function dispose<T extends IDisposable>(arg: T | IterableIterator<T> | un
 
 		return Array.isArray(arg) ? [] : arg;
 	} else if (arg) {
-		markTracked(arg);
 		arg.dispose();
 		return arg;
 	}
 }
 
 
 export function combinedDisposable(...disposables: IDisposable[]): IDisposable {
-	disposables.forEach(markTracked);
-	return toDisposable(() => dispose(disposables));
+	const parent = toDisposable(() => dispose(disposables));
+	setParentOfDisposables(disposables, parent);
+	return parent;
 }
 
 export function toDisposable(fn: () => void): IDisposable {
 	const self = trackDisposable({
 		dispose: () => {
-			markTracked(self);
+			markAsDisposed(self);
+			fn();
+		}
+	});
+	return self;
+}
+
+function noop(): void { }
+
+/**
+ * Creates a disposable that calls `fn` and then clears its reference to `fn` when being disposed.
+ * This prevents memory leaks and ensures `fn` is called at most once.
+*/
+export function toNonLeakingDisposable(fn: () => void): IDisposable {
```

Comment:
Any particular reason why `toDisposable` should not do the same `self.dispose = noop` ?

Reply:
That would change the semantics of `toDisposable`: Currently, multiple calls of `dispose` will lead to multiple calls of `fn`.
With `self.dispose = noop`, multiple calls of `dispose` will only call `fn` once.

I don't know if anyone depends on this behavior.

Reply:
cc @jrieken @mjbvz That seems a wrong implementation of `toDisposable`. Could we try fixing that in the debt week (even outside this PR) to maximize our selfhosting coverage?

Reply:
I think @joaomoreno is Mr `toDisposable` but I agree that this should be called only once

Reply:
Tests are all happy, optimistically pushed that: https://github.com/microsoft/vscode/commit/f8cf8f39be4dbf7bcab2a22181ab571ee40d2b6d

Reply:
> I think @joaomoreno is Mr `toDisposable` but I agree that this should be called only once

As if _multiple calls of `dispose` will lead to multiple calls of `fn`_ would have any reason other than _didn't think about that 🤷‍♂️_.

Let's have `toDisposable` do exactly what `toNonLeakingDisposable` does.